### PR TITLE
Implement Regolith deposit tx gasUsed changes

### PR DESCRIPTION
**Pull request by ajsutton**
**Description**

Actual gas used is now reported in deposit tx receipts rather than always being 0 or the tx gas limit. Changes are all directly behind checks for regolith being enabled so bedrock compatibility is preserved.

One key thing to note is that for the gasUsed to be accurate, gas refunds from things like self-destruct or clearing storage slots must be applied. This is done in the `st.refundGas` call which is completely skipped in Bedrock.  `refundGas` _also_ refunds the ETH that was paid upfront for the unused gas. Since deposit tx hard code the gas price to 0 however, the ETH refunded is always 0 (because gasPrice * unusedGas will always be 0). So in Regolith we allow `refundGas` to be called so the gas used accounting is correct and then bail out before the payments to coinbase (which wind up being negative because of the 0 gasFeeCap).

**Tests**

No tests at op-geth level as this is very hard to unit test. https://github.com/ethereum-optimism/optimism/pull/4931 will add e2e tests that cover this (including verifying that unused gas doesn't result in an ETH payment to the sender).

Follows on from https://github.com/ethereum-optimism/op-geth/pull/51

**Metadata**
- Fixes https://linear.app/optimism/issue/CLI-3397/op-geth-regolith-gasused-changed

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -336,7 +336,7 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 		// Record deposits as using all their gas (matches the gas pool)
 		// System Transactions are special & are not recorded as using any gas (anywhere)
 		gasUsed := st.msg.Gas()
-		if st.msg.IsSystemTx() {
+		if st.msg.IsSystemTx() && !st.evm.ChainConfig().IsRegolith(st.evm.Context.Time) {
```
Comment (protolambda) [inline core/state_transition.go:339]: Needs a comment update about the regolith tweak.

File: core/state_transition.go [hunk 2]
```diff
@@ -417,7 +417,7 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 	}
 
 	// if deposit: skip refunds, skip tipping coinbase
-	if st.msg.IsDepositTx() {
+	if st.msg.IsDepositTx() && !rules.IsOptimismRegolith {
```
Comment (protolambda) [inline core/state_transition.go:420]: Needs a comment about regolith tweak

**Review by protolambda — APPROVED** [review]
LGTM, but would update some comments.

FWIW, the self-destruct refund isn't applicable anymore after https://eips.ethereum.org/EIPS/eip-3529 but the sstore one is (although reduced), so it still makes sense call the refund function and not do the fee-recipient, nice catch.

File: core/state_transition.go (hunk 1 above)
Comment (ajsutton) [reply core/state_transition.go:341]: Added comments - let me know if they fit what you expect.

**Pull request merged**